File: tests/algorithms/x/PHP/maths/modular_division.php
<?php
ini_set('memory_limit', '-1');
$now_seed = 0;
$now_seeded = false;
$s = getenv('MOCHI_NOW_SEED');
if ($s !== false && $s !== '') {
    $now_seed = intval($s);
    $now_seeded = true;
}
function _now() {
    global $now_seed, $now_seeded;
    if ($now_seeded) {
        $now_seed = ($now_seed * 1664525 + 1013904223) % 2147483647;
        return $now_seed;
    }
    return hrtime(true);
}
function _str($x) {
    if (is_array($x)) {
        $isList = array_keys($x) === range(0, count($x) - 1);
        if ($isList) {
            $parts = [];
            foreach ($x as $v) { $parts[] = _str($v); }
            return '[' . implode(' ', $parts) . ']';
        }
        $parts = [];
        foreach ($x as $k => $v) { $parts[] = _str($k) . ':' . _str($v); }
        return 'map[' . implode(' ', $parts) . ']';
    }
    if (is_bool($x)) return $x ? 'true' : 'false';
    if ($x === null) return 'null';
    return strval($x);
}
function _intdiv($a, $b) {
    if (function_exists('bcdiv')) {
        $sa = is_int($a) ? strval($a) : (is_string($a) ? $a : sprintf('%.0f', $a));
        $sb = is_int($b) ? strval($b) : (is_string($b) ? $b : sprintf('%.0f', $b));
        return intval(bcdiv($sa, $sb, 0));
    }
    return intdiv($a, $b);
}
$__start_mem = memory_get_usage();
$__start = _now();
  function mod($a, $n) {
  $r = $a % $n;
  if ($r < 0) {
  return $r + $n;
}
  return $r;
};
  function greatest_common_divisor($a, $b) {
  $x = ($a < 0 ? -$a : $a);
  $y = ($b < 0 ? -$b : $b);
  while ($y != 0) {
  $t = $x % $y;
  $x = $y;
  $y = $t;
};
  return $x;
};
  function extended_gcd($a, $b) {
  if ($b == 0) {
  return [$a, 1, 0];
}
  $res = extended_gcd($b, $a % $b);
  $d = $res[0];
  $p = $res[1];
  $q = $res[2];
  $x = $q;
  $y = $p - $q * (_intdiv($a, $b));
  return [$d, $x, $y];
};
  function extended_euclid($a, $b) {
  if ($b == 0) {
  return [1, 0];
}
  $res = extended_euclid($b, $a % $b);
  $x = $res[1];
  $y = $res[0] - (_intdiv($a, $b)) * $res[1];
  return [$x, $y];
};
  function invert_modulo($a, $n) {
  $res = extended_euclid($a, $n);
  $inv = $res[0];
  return mod($inv, $n);
};
  function modular_division($a, $b, $n) {
  if ($n <= 1) {
  $panic('n must be > 1');
}
  if ($a <= 0) {
  $panic('a must be > 0');
}
  if (greatest_common_divisor($a, $n) != 1) {
  $panic('gcd(a,n) != 1');
}
  $eg = extended_gcd($n, $a);
  $s = $eg[2];
  return mod($b * $s, $n);
};
  function modular_division2($a, $b, $n) {
  $s = invert_modulo($a, $n);
  return mod($b * $s, $n);
};
  function tests() {
  if (modular_division(4, 8, 5) != 2) {
  $panic('md1');
}
  if (modular_division(3, 8, 5) != 1) {
  $panic('md2');
}
  if (modular_division(4, 11, 5) != 4) {
  $panic('md3');
}
  if (modular_division2(4, 8, 5) != 2) {
  $panic('md21');
}
  if (modular_division2(3, 8, 5) != 1) {
  $panic('md22');
}
  if (modular_division2(4, 11, 5) != 4) {
  $panic('md23');
}
  if (invert_modulo(2, 5) != 3) {
  $panic('inv');
}
  $eg = extended_gcd(10, 6);
  if ($eg[0] != 2 || $eg[1] != (-1) || $eg[2] != 2) {
  $panic('eg');
}
  $eu = extended_euclid(10, 6);
  if ($eu[0] != (-1) || $eu[1] != 2) {
  $panic('eu');
}
  if (greatest_common_divisor(121, 11) != 11) {
  $panic('gcd');
}
};
  function main() {
  tests();
  echo rtrim(_str(modular_division(4, 8, 5))), PHP_EOL;
};
  main();
$__end = _now();
$__end_mem = memory_get_peak_usage();
$__duration = max(1, intdiv($__end - $__start, 1000));
$__mem_diff = max(0, $__end_mem - $__start_mem);
$__bench = ["duration_us" => $__duration, "memory_bytes" => $__mem_diff, "name" => "main"];
$__j = json_encode($__bench, 128);
$__j = str_replace("    ", "  ", $__j);
echo $__j, PHP_EOL;

File: tests/algorithms/transpiler/FS/maths/modular_division.fs
// Generated 2025-08-08 18:09 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let _dictAdd<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) (v:'V) =
    d.[k] <- v
    d
let _dictCreate<'K,'V when 'K : equality> (pairs:('K * 'V) list) : System.Collections.Generic.IDictionary<'K,'V> =
    let d = System.Collections.Generic.Dictionary<'K, 'V>()
    for (k, v) in pairs do
        d.[k] <- v
    upcast d
let _dictGet<'K,'V when 'K : equality> (d:System.Collections.Generic.IDictionary<'K,'V>) (k:'K) : 'V =
    match d.TryGetValue(k) with
    | true, v -> v
    | _ -> Unchecked.defaultof<'V>
let _idx (arr:'a array) (i:int) : 'a =
    if not (obj.ReferenceEquals(arr, null)) && i >= 0 && i < arr.Length then arr.[i] else Unchecked.defaultof<'a>
let rec _str v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
let _floordiv (a:int) (b:int) : int =
    let q = a / b
    let r = a % b
    if r <> 0 && ((a < 0) <> (b < 0)) then q - 1 else q
let rec ``mod`` (a: int) (n: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable a = a
    let mutable n = n
    try
        let r: int = ((a % n + n) % n)
        if r < 0 then
            __ret <- r + n
            raise Return
        __ret <- r
        raise Return
        __ret
    with
        | Return -> __ret
and greatest_common_divisor (a: int) (b: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable a = a
    let mutable b = b
    try
        let mutable x: int = if a < 0 then (-a) else a
        let mutable y: int = if b < 0 then (-b) else b
        while y <> 0 do
            let t: int = ((x % y + y) % y)
            x <- y
            y <- t
        __ret <- x
        raise Return
        __ret
    with
        | Return -> __ret
and extended_gcd (a: int) (b: int) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable a = a
    let mutable b = b
    try
        if b = 0 then
            __ret <- unbox<int array> [|a; 1; 0|]
            raise Return
        let res: int array = extended_gcd (b) (((a % b + b) % b))
        let d: int = _idx res (0)
        let p: int = _idx res (1)
        let q: int = _idx res (2)
        let mutable x: int = q
        let mutable y: int64 = (int64 p) - ((int64 q) * (int64 (_floordiv a b)))
        __ret <- Array.map int [|int64 (d); int64 (x); y|]
        raise Return
        __ret
    with
        | Return -> __ret
and extended_euclid (a: int) (b: int) =
    let mutable __ret : int array = Unchecked.defaultof<int array>
    let mutable a = a
    let mutable b = b
    try
        if b = 0 then
            __ret <- unbox<int array> [|1; 0|]
            raise Return
        let res: int array = extended_euclid (b) (((a % b + b) % b))
        let mutable x: int = _idx res (1)
        let mutable y: int64 = (int64 (_idx res (0))) - ((int64 (_floordiv a b)) * (int64 (_idx res (1))))
        __ret <- Array.map int [|int64 (x); y|]
        raise Return
        __ret
    with
        | Return -> __ret
and invert_modulo (a: int) (n: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable a = a
    let mutable n = n
    try
        let res: int array = extended_euclid (a) (n)
        let inv: int = _idx res (0)
        __ret <- int (``mod`` (inv) (n))
        raise Return
        __ret
    with
        | Return -> __ret
and modular_division (a: int) (b: int) (n: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable a = a
    let mutable b = b
    let mutable n = n
    try
        if n <= 1 then
            failwith ("n must be > 1")
        if a <= 0 then
            failwith ("a must be > 0")
        if (greatest_common_divisor (a) (n)) <> 1 then
            failwith ("gcd(a,n) != 1")
        let eg: int array = extended_gcd (n) (a)
        let s: int = _idx eg (2)
        __ret <- int (``mod`` (int ((int64 b) * (int64 s))) (n))
        raise Return
        __ret
    with
        | Return -> __ret
and modular_division2 (a: int) (b: int) (n: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable a = a
    let mutable b = b
    let mutable n = n
    try
        let s: int = invert_modulo (a) (n)
        __ret <- int (``mod`` (int ((int64 b) * (int64 s))) (n))
        raise Return
        __ret
    with
        | Return -> __ret
and tests () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        if (modular_division (4) (8) (5)) <> 2 then
            failwith ("md1")
        if (modular_division (3) (8) (5)) <> 1 then
            failwith ("md2")
        if (modular_division (4) (11) (5)) <> 4 then
            failwith ("md3")
        if (modular_division2 (4) (8) (5)) <> 2 then
            failwith ("md21")
        if (modular_division2 (3) (8) (5)) <> 1 then
            failwith ("md22")
        if (modular_division2 (4) (11) (5)) <> 4 then
            failwith ("md23")
        if (invert_modulo (2) (5)) <> 3 then
            failwith ("inv")
        let eg: int array = extended_gcd (10) (6)
        if (((_idx eg (0)) <> 2) || ((_idx eg (1)) <> (-1))) || ((_idx eg (2)) <> 2) then
            failwith ("eg")
        let eu: int array = extended_euclid (10) (6)
        if ((_idx eu (0)) <> (-1)) || ((_idx eu (1)) <> 2) then
            failwith ("eu")
        if (greatest_common_divisor (121) (11)) <> 11 then
            failwith ("gcd")
        __ret
    with
        | Return -> __ret
and main () =
    let mutable __ret : unit = Unchecked.defaultof<unit>
    try
        let __bench_start = _now()
        let __mem_start = System.GC.GetTotalMemory(true)
        tests()
        printfn "%s" (_str (modular_division (4) (8) (5)))
        let __bench_end = _now()
        let __mem_end = System.GC.GetTotalMemory(true)
        printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

        __ret
    with
        | Return -> __ret
main()
